src/lib.rs

```diff
@@ -20,29 +20,32 @@ pub use backend::{
 pub use bool::*;
 pub use ntt::{Ntt, NttBackendU64, NttInit};
 #[cfg(any(feature = "interactive_mp", feature = "non_interactive_mp"))]
-pub use shortint::{div_zero_error_flag, reset_error_flags, FheUint8};
-pub use shortint::SeededBatchedFheUint8;
-
+pub use shortint::{
+    div_zero_error_flag, reset_error_flags, EncFheUint8, FheUint8, SeededBatchedFheUint8,
+};
 
 pub use decomposer::{Decomposer, DecomposerIter, DefaultDecomposer};
 
+type RowIterator<'a, T> = Box<dyn Iterator<Item = &'a T> + 'a>;
+type RowIteratorMut<'a, T> = Box<dyn Iterator<Item = &'a mut T> + 'a>;
+
 pub trait Matrix: AsRef<[Self::R]> {
     type MatElement;
     type R: Row<Element = Self::MatElement>;
 
     /// (Rows, Cols)
     fn dimension(&self) -> (usize, usize);
 
-    fn get_row(&self, row_idx: usize) -> impl Iterator<Item = &Self::MatElement> {
-        self.as_ref()[row_idx].as_ref().iter().map(move |r| r)
+    fn get_row(&self, row_idx: usize) -> RowIterator<'_, Self::MatElement> {
+        Box::new(self.as_ref()[row_idx].as_ref().iter().map(move |r| r))
     }
 
     fn get_row_slice(&self, row_idx: usize) -> &[Self::MatElement] {
         self.as_ref()[row_idx].as_ref()
     }
 
-    fn iter_rows(&self) -> impl Iterator<Item = &Self::R> {
-        self.as_ref().iter().map(move |r| r)
+    fn iter_rows(&self) -> RowIterator<'_, Self::R> {
+        Box::new(self.as_ref().iter().map(move |r| r))
```

Comment:
Can you share reasons on why adding boxes in get_row, iter_rows, and iter_rows_mut?

Reply:
oh yeah, for some reason i'm not able to build without doing this. I'm getting some error that `impl Iterator` isn't available in my version of Rust?

Reply:
are you guys using nightly rust? or what version? @ChihChengLiang